numerics/quadrature_body.hpp

```diff
@@ -80,65 +158,57 @@ AutomaticClenshawCurtisImplementation(
+          cached_f_cos_N⁻¹π);
     }
   }
   return estimate;
 }
 
-template<int initial_points, typename Argument, typename Function>
+template<int points, typename Argument, typename Function>
 Primitive<std::invoke_result_t<Function, Argument>, Argument>
-AutomaticClenshawCurtis(
+ClenshawCurtisImplementation(
     Function const& f,
     Argument const& lower_bound,
     Argument const& upper_bound,
-    std::optional<double> const max_relative_error,
-    std::optional<int> const max_points) {
-  using Result = Primitive<std::invoke_result_t<Function, Argument>, Argument>;
-  // TODO(egg): factor the evaluations of f.
-  Result const estimate =
-      ClenshawCurtis<initial_points>(f, lower_bound, upper_bound);
-  return AutomaticClenshawCurtisImplementation<
-      initial_points + initial_points - 1>(
-      f, lower_bound, upper_bound, max_relative_error, max_points, estimate);
-}
-
-template<int points, typename Argument, typename Function>
-Primitive<std::invoke_result_t<Function, Argument>, Argument> ClenshawCurtis(
-    Function const& f,
-    Argument const& lower_bound,
-    Argument const& upper_bound) {
+    std::vector<std::invoke_result_t<Function, Argument>>& cached_f_cos_N⁻¹π) {
   // We follow the notation from [Gen72b] and [Gen72c].
   using Value = std::invoke_result_t<Function, Argument>;
 
   constexpr int N = points - 1;
   constexpr int log2_N = FloorLog2(N);
-  static_assert(N == 1 << log2_N);
 
   Difference<Argument> const half_width = (upper_bound - lower_bound) / 2;
-
   constexpr Angle N⁻¹π = π * Radian / N;
 
-  // We use a discrete Fourier transform rather than a cosine transform, see
-  // [Gen72c], equation (3).
-  // TODO(egg): Consider a discrete cosine transform, ideally incrementally
-  // computed to improve the performance of the automatic quadrature.
+  FillClenshawCurtisCache<points>(
+      f, lower_bound, upper_bound, cached_f_cos_N⁻¹π);
 
-  // If we identify [lower_bound, upper_bound] with [-1, 1],
-  // f_cos_N⁻¹π[s] is f(cos πs/N).
+  // TODO(phl): If might be possible to avoid copies since cached_f_cos_N⁻¹π is
+  // tantalizing close to the order needed for the FFT.
   std::vector<Value> f_cos_N⁻¹π;
   f_cos_N⁻¹π.resize(2 * N);
-  for (int s = 0; s <= N; ++s) {
-    // The N + 1 evaluations.
-    f_cos_N⁻¹π[s] = f(lower_bound + half_width * (1 + Cos(N⁻¹π * s)));
-  }
-  for (int s = N + 1; s <= 2 * N - 1; ++s) {
-    f_cos_N⁻¹π[s] = f_cos_N⁻¹π[2 * N - s];  // [Gen72c] (5).
+  // An index in f_cos_N⁻¹π, corresponding to the increasing order of s for this
+  // value of N.
+  int reverse = 0;
+  // An index in cached_f_cos_N⁻¹π, corresponding to the order in which the
+  // values were put in the cache.  Note that entry 0 is special and holds the
+  // value corresponding to s = N.
+  for (int direct = 1; direct <= N; ++direct) {
+    f_cos_N⁻¹π[reverse] = cached_f_cos_N⁻¹π[direct];
+    if (reverse > 0) {
+      f_cos_N⁻¹π[2 * N - reverse] = cached_f_cos_N⁻¹π[direct];  // [Gen72c] (5).
```

Comment:
Call reverse `s` and write `f_cos_N⁻¹π[2 * N - s] = f_cos_N⁻¹π[s]` so that the reference to the paper still makes sense.